ldap/servers/slapd/slapi-memberof.c

```diff
@@ -0,0 +1,1280 @@
+        }
+        slapi_ch_free((void **)&cookie);
+        slapi_ch_free_string(&filter_str);
+    }
+
+    if (free_it) {
+        slapi_ch_free_string(&escaped_filter_val);
+    }
+    return rc;
+}
+
+
+static int
+Xmemberof_get_groups_r(Slapi_MemberOfConfig *config, Slapi_DN *member_sdn, Xmemberof_get_groups_data *data)
+{
+    Slapi_ValueSet *groupvals = slapi_valueset_new();
+    Slapi_ValueSet *group_norm_vals = slapi_valueset_new();
+    Slapi_ValueSet *nsuniqueidvals = slapi_valueset_new();
+    Slapi_Value *member_ndn_val =
+        slapi_value_new_string(slapi_sdn_get_ndn(member_sdn));
+    int rc;
+    int cached = 0;
+
+    slapi_value_set_flags(member_ndn_val, SLAPI_ATTR_FLAG_NORMALIZED_CIS);
+
+    Xmemberof_get_groups_data member_data = {config, member_ndn_val, &groupvals, &group_norm_vals, &nsuniqueidvals, data->already_seen_ndn_vals, data->use_cache};
+
+/* Search for any grouping attributes that point to memberdn.
+     * For each match, add it to the list, recurse and do same search */
+#if MEMBEROF_CACHE_DEBUG
+    slapi_log_err(SLAPI_LOG_ERR, "linked_attributes", "memberof_get_groups_r: Ancestors of %s\n", slapi_sdn_get_dn(member_sdn));
+#endif
+    rc = Xmemberof_call_foreach_dn(NULL, member_sdn, config, config->groupattrs,
+                                  Xmemberof_get_groups_callback, &member_data, &cached, member_data.use_cache);
+
+    Xmerge_ancestors(&member_ndn_val, &member_data, data);
+    if (!cached && member_data.use_cache)
+        Xcache_ancestors(config, &member_ndn_val, &member_data);
+
+    slapi_value_free(&member_ndn_val);
+    slapi_valueset_free(groupvals);
+    slapi_valueset_free(group_norm_vals);
+    slapi_valueset_free(nsuniqueidvals);
+    
+
+    return rc;
+}
+
+static PRIntn Xmemberof_hash_compare_keys(const void *v1, const void *v2)
+{
+    PRIntn rc;
+    if (0 == strcasecmp((const char *) v1, (const char *) v2)) {
+        rc = 1;
+    } else {
+        rc = 0;
+    }
+    return rc;
+}
+
+static PRIntn Xmemberof_hash_compare_values(const void *v1, const void *v2)
+{
+    PRIntn rc;
+    if ((char *) v1 == (char *) v2) {
+        rc = 1;
+    } else {
+        rc = 0;
+    }
+    return rc;
+}
+
+/*
+ *  Hashing function using Bernstein's method
+ */
+static PLHashNumber Xmemberof_hash_fn(const void *key)
+{
+    PLHashNumber hash = 5381;
+    unsigned char *x = (unsigned char *)key;
+    int c;
+
+    while ((c = *x++)){
+        hash = ((hash << 5) + hash) ^ c;
+    }
+    return hash;
+}
+
+/* allocates the plugin hashtable
+ * This hash table is used by operation and is protected from
+ * concurrent operations with the memberof_lock (if not usetxn, memberof_lock
+ * is not implemented and the hash table will be not used.
+ *
+ * The hash table contains all the DN of the entries for which the memberof
+ * attribute has been computed/updated during the current operation
+ *
+ * hash table should be empty at the beginning and end of the plugin callback
```

Comment:
Do we actually assert that anywhere?

Reply:
No assert. The hash table is allocated at the beginning  of slapi_memberof and then emptied (Xancestor_hashtable_empty)/destroy at the end.

Reply:
It would be good to have some PR_ASSERTS that validate that assumption then during tests IMO.